src/containers/projects-table/projects-table.tsx

```diff
@@ -3,54 +3,48 @@
  */
 
-import { useEffect, useState } from 'react';
-import { AxiosResponse } from 'axios';
 import { Project, WorkPackage } from 'utils';
-import { apiFetch } from '../../shared/axios';
 import { weeksPipe, fullNamePipe, wbsPipe } from '../../shared/pipes';
 import PrjsTable from '../../components/projects-table/projects-table'; // Directly rename the default import
-import { DisplayProject } from '../../components/projects-table/projects-table';
 import './projects-table.module.css';
+import { useAllProjects } from '../../services/projects';
 
 const ProjectsTable: React.FC = () => {
-  const [allProjects, setAllProjects] = useState<DisplayProject[]>([]); // store projects data
-
-  // Transforms given project data and sets local state
-  const updateData = (response: AxiosResponse) => {
-    setAllProjects(
-      response.data.map((prj: Project) => {
-        return {
-          wbsNum: wbsPipe(prj.wbsNum),
-          name: prj.name,
-          projectLead: fullNamePipe(prj.projectLead),
-          projectManager: fullNamePipe(prj.projectManager),
-          duration: weeksPipe(
-            prj.workPackages.reduce((tot: number, cur: WorkPackage) => tot + cur.duration, 0)
-          )
-        };
-      })
-    );
+  const { isLoading, errorMessage, responseData } = useAllProjects();
+
+  /**
+   * Transform given list of projects into a list of display projects.
+   * @param projects list of projects to transform
+   * @returns list of display projects
+   */
+  const updateData = (projects: Project[]) => {
+    return projects.map((prj: Project) => {
+      return {
+        wbsNum: wbsPipe(prj.wbsNum),
+        name: prj.name,
+        projectLead: fullNamePipe(prj.projectLead),
+        projectManager: fullNamePipe(prj.projectManager),
+        duration: weeksPipe(
+          prj.workPackages.reduce((tot: number, cur: WorkPackage) => tot + cur.duration, 0)
+        )
+      };
+    });
   };
 
-  // Fetch list of projects from API on component loading
-  useEffect(() => {
-    let mounted = true; // indicates component is mounted
-
-    const fetchProjects = async () => {
-      apiFetch
-        .get('/projects')
-        .then((response: AxiosResponse) => (mounted ? updateData(response) : ''))
-        .catch((error) => (mounted ? console.log('fetch projects error: ' + error.message) : ''));
-    };
-    fetchProjects();
-
-    // cleanup function indicates component has been unmounted
-    return () => {
-      mounted = false;
-    };
-  }, []);
+  if (isLoading) {
+    return <p>Loading...</p>;
+  }
+  if (errorMessage !== '' || responseData === undefined) {
+    return (
+      <>
+        <h3>Oops, sorry!</h3>
+        <h5>There was an error loading the page.</h5>
+        <p>{errorMessage ? errorMessage : 'The data did not load properly.'}</p>
+      </>
+    );
+  }
 
-  return <PrjsTable allProjects={allProjects} />;
+  return <PrjsTable allProjects={updateData(responseData)!} />;
```

Comment:
Any idea why the ! is on the right side of the closing parentheses? I'm not quite familiar enough / maybe not remembering. Does it not work putting the ! to the left of the closing parentheses?

Reply:
I believe including the ! on the right side is telling the compiler that the return value of the function call is non-null whereas including it on the left side just means that responseData is non-null. I honestly hadn't seen this syntax before so I had to do some googling.